crates/crypto_api/src/crypto_system/mod.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{Buffer, CryptoResult};
+
+/// A trait describing a cryptographic system implementation compatible
+/// with Lib3h and Holochain.
+#[allow(clippy::borrowed_box)]
+pub trait CryptoSystem: Sync {
+    /// Crypto System is designed to be used as a trait-object
+    /// Since we can't get a sized clone, provide clone in a Box.
+    fn box_clone(&self) -> Box<dyn CryptoSystem>;
+
+    /// helps work around some sizing issues with rust trait-objects
+    fn as_crypto_system(&self) -> &dyn CryptoSystem;
+
+    /// create a new memory secured buffer
+    /// that is compatible with this crypto system
+    fn buf_new_secure(&self, size: usize) -> Box<dyn Buffer>;
+
+    /// this is just a helper to create a
+    /// sized boxed Vec<u8> as a Box<dyn Buffer>
+    fn buf_new_insecure(&self, size: usize) -> Box<dyn Buffer> {
+        Box::new(vec![0; size])
+    }
+
+    // -- random methods -- //
+
+    /// fill all the bytes in the buffer with secure random data
+    fn randombytes_buf(&self, buffer: &mut Box<dyn Buffer>) -> CryptoResult<()>;
+
+    // -- signature methods -- //
+
+    /// bytelength of signature seed
+    fn sign_seed_bytes(&self) -> usize;
+
+    /// bytelength of signature public key
+    fn sign_public_key_bytes(&self) -> usize;
+
+    /// bytelength of signature secret key
+    fn sign_secret_key_bytes(&self) -> usize;
+
+    /// bytelength of a digital signature
+    fn sign_bytes(&self) -> usize;
+
+    /// generate a deterministic signature public / secret keypair
+    /// based off the given seed entropy
+    fn sign_seed_keypair(
+        &self,
+        seed: &Box<dyn Buffer>,
+        public_key: &mut Box<dyn Buffer>,
+        secret_key: &mut Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// generate a pure entropy based signature public / secret keypair
+    fn sign_keypair(
+        &self,
+        public_key: &mut Box<dyn Buffer>,
+        secret_key: &mut Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// generate a digital signature for `message` with the given secret key
+    fn sign(
+        &self,
+        signature: &mut Box<dyn Buffer>,
+        message: &Box<dyn Buffer>,
+        secret_key: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
+
+    /// verify that the digital `signature` is valid for given `message` and
+    /// `public_key`
+    fn sign_verify(
+        &self,
+        signature: &Box<dyn Buffer>,
+        message: &Box<dyn Buffer>,
+        public_key: &Box<dyn Buffer>,
+    ) -> CryptoResult<bool>;
+}
+
+pub mod crypto_system_test;
```

Comment:
I believe you can put this under `#[cfg(test)]` so it compiles only during `cargo test` .

Reply:
I tried that to start, but then the lib3h_sodium crate doesn't have access to it (even in tests)... If anyone knows a better way to organize this so the test code doesn't get build/exposed in the production lib, I'd love the feedback.